sky/data/storage.py

```diff
@@ -3547,3 +3607,450 @@ def _delete_cos_bucket(self):
+
+    def batch_oci_rsync(self,
+                        source_path_list: List[Path],
+                        create_dirs: bool = False) -> None:
+        """Invokes oci sync to batch upload a list of local paths to Bucket
+
+        Use OCI bulk operation to batch process the file upload
+
+        Args:
+            source_path_list: List of paths to local files or directories
+            create_dirs: If the local_path is a directory and this is set to
+                False, the contents of the directory are directly uploaded to
+                root of the bucket. If the local_path is a directory and this is
+                set to True, the directory is created in the bucket root and
+                contents are uploaded to it.
+        """
+
+        def get_file_sync_command(base_dir_path, file_names):
+            includes = ' '.join(
+                [f'--include "{file_name}"' for file_name in file_names])
+            sync_command = (
+                'oci os object bulk-upload --no-follow-symlinks --overwrite '
+                f'--bucket-name {self.name} --namespace-name {self.namespace} '
+                f'--src-dir "{base_dir_path}" {includes}')
+
+            all_commands = oci.goto_oci_cli_venv()
+            all_commands.append(sync_command)
+            all_commands.append(oci.leave_oci_cli_venv())
+
+            return ' && '.join(all_commands)
+
+        def get_dir_sync_command(src_dir_path, dest_dir_name):
+            if dest_dir_name and not str(dest_dir_name).endswith('/'):
+                dest_dir_name = f'{dest_dir_name}/'
+
+            # we exclude .git directory from the sync
+            sync_command = (
+                'oci os object bulk-upload --no-follow-symlinks --overwrite '
+                f'--bucket-name {self.name} --namespace-name {self.namespace} '
+                f'--object-prefix "{dest_dir_name}" --src-dir "{src_dir_path}" '
+                f'--exclude ".git/*" ')
+
+            all_commands = oci.goto_oci_cli_venv()
+            all_commands.append(sync_command)
+            all_commands.append(oci.leave_oci_cli_venv())
+
+            return ' && '.join(all_commands)
+
+        # Generate message for upload
+        if len(source_path_list) > 1:
+            source_message = f'{len(source_path_list)} paths'
+        else:
+            source_message = source_path_list[0]
+
+        with rich_utils.safe_status(
+                f'[bold cyan]Syncing '
+                f'[green]{source_message}[/] to [green]oci://{self.name}/[/]'):
+            data_utils.parallel_upload(source_path_list,
+                                       get_file_sync_command,
+                                       get_dir_sync_command,
+                                       self.name,
+                                       self._ACCESS_DENIED_MESSAGE,
+                                       create_dirs=create_dirs,
+                                       max_concurrent_uploads=1)
```

Comment:
Why not set it to this constant?

https://github.com/skypilot-org/skypilot/blob/2bd7c3ed35ce27d5ffeb9010e23da8d9ebb3ffa7/sky/data/storage.py#L61

Reply:
We use OCI bulk-upload to parallel upload instead of concurrent threads count here.

Reply:
It will be great to add a comment with this :))